docs/design/downward_api_resources_limits_requests.md

```diff
@@ -0,0 +1,530 @@
+Exposing the requested and limited amounts of available resources inside containers will allow
+these applications to be configured more easily. Although docker already
+exposes some of this information inside containers, the downward API helps
+exposing this information in a runtime-agnostic manner in Kubernetes.
+
+## Design
+
+This is mostly driven by the discussion in [this issue](https://github.com/kubernetes/kubernetes/issues/9473).
+There are three approaches discussed in this document to obtain resources limits
+and requests to be exposed as environment variables and volumes inside
+containers:
+
+1. The first approach requires users to specify full json path selectors
+in which selectors are relative to the pod spec. The benefit of this
+approach is to specify pod level resources, and since containers are
+also part of a pod spec, it can be used to specify container level
+resources too.
+
+2. The second approach requires specifying partial json path selectors
+which are relative to the container spec. This approach helps
+in retrieving a container specific resource limits and requests, and at
+the same time, it is simpler to specify than full json path selectors.
+
+3. In this approach, users specify particular strings to retrieve
+resources limits and requests and do not specify any json path
+selectors. This approach is similar to the existing downward API
+implementation approach. The advantages of this approach is that it is
+simpler to specify that the firs two, and does not require any type of
+conversion between internal and versioned objects or json selectors as
+discussed below. However, unlike the full and partial path selectors
+approaches can be extended to other resources with no or little changes,
+this approach might require a relatively more work to extend them to
+other resources in future if needed.
+
+Before discussing a bit more about merits of each approach, here is a
+brief discussion about json path selectors and some implications related
+to their use.
+
+#### JSONpath selectors
+
+Versioned objects in kubernetes have json tags as part of their golang fields. Although currently internal objects
+in kubernetes also have json tags but these tags should be removed in
+future (see [3933](https://github.com/kubernetes/kubernetes/issues/3933)
+for discussion). So for discussion in this proposal, we assume that
+internal objects do not have json tags. In the first two approaches
+(full and partial json selectors), when a user creates a pod and its
+containers, the user specifies a json path selector in the pod's
+spec to retrieve values of its limits and requests. The selector
+is composed of json tags similar to json paths used with kubectl
+([json](http://kubernetes.io/docs/user-guide/jsonpath/)). This proposal
+uses kubernetes' json path library to process the selectors to retrieve
+the values. As kubelet operates on internal objects (without json tags),
```

Comment:
@lavalamp, for this work (targeted for 1.3), if we use json path selectors for accessing resource values in kubelet, we would need to convert internal objects to versioned objects unless kubelet starts using external versions. If we are ok with that then not a blocker. Also not a blocker, if we just go with no selectors (or magic keys) approach.